src/invert/laplace/impls/cyclic/cyclic_laplace.cxx

```diff
@@ -90,7 +90,7 @@ const FieldPerp LaplaceCyclic::solve(const FieldPerp &rhs, const FieldPerp &x0)
   FieldPerp x(mesh); // Result
   x.allocate();
 
-  Coordinates *coord = mesh->coordinates();
+  Coordinates *coord = mesh->coordinates(location);
```

Comment:
One thing I noticed is that we don't force/check that the coefficients are set at a consistent location so I think we might need a bit more work to make this a bit more robust?

Are you aware of any implementations that do cache location sensitive values in the constructor at the moment?

Reply:
If FieldPerp had a variable location I guess this would help in these efforts -- can get location from passed fields and if needed check this is consistent with any precalculated data.

Reply:
(I found `grep -i ">coordinates(" $(find ./{src,include} -name '*.?xx')` quite useful in finding any uses of coordinates, from which it's easy to pick out the few places where a location is not passed).

Reply:
>Are you aware of any implementations that do cache location sensitive values in the constructor at the moment?

No, not at the moment.

Reply:
> One thing I noticed is that we don't force/check that the coefficients are set at a consistent location so I think we might need a bit more work to make this a bit more robust?

That might be a reason for requiring the location in the constructor? If we know the location is set then we can check the location of the coefficients when they are passed in. Without having a location fixed from the constructor I can't think how we could guarantee a consistent location between calls to `setCoef*()` and `solve()`?

Reply:
To remove dependence on the global `mesh`, should we (optionally) pass a mesh to the Laplacian constructor as well, then check that coefficients and rhs use the same mesh as the Laplacian?

Reply:
> Without having a location fixed from the constructor I can't think how we could guarantee a consistent location between calls to setCoef*() and solve()?

I guess one possibility would be to set the location of the Laplacian object from the first coefficient or rhs passed in, and then check the others against that. I'm not in favour of that solution because it adds more code to go wrong (have to keep track of whether location has been set). I'd favour keeping the simple, explicit solution of setting the location of the Laplacian object when it's created as it is now.